waspc/packages/deploy/src/providers/fly/cmd/cmd.ts

```diff
@@ -0,0 +1,48 @@
+import { $, cd } from 'zx'
+import { ClientCommonOps, CommonOps, ServerCommonOps } from '../helpers/CommonOps.js'
+import { buildDirExists, getCommandHelp, waspSays } from '../helpers/helpers.js'
+import * as tomlHelpers from '../helpers/tomlFileHelpers.js'
+import { executeFlyCommand } from '../index.js'
+import { CmdOptions, SERVER_CONTEXT_OPTION } from './CmdOptions.js'
+
+// Runs a command by copying down the project toml files, executing it, and copying it back up (just in case).
+// If the toml file does not exist, some commands will not run with additional args (e.g. -a <appname>).
+export async function cmd(flyctlArgs: string[], options: CmdOptions) {
+  waspSays(`Running ${options.context} command: flyctl ${flyctlArgs.join(' ')}`)
+
+  if (!buildDirExists(options.waspDir)) {
+    waspSays('Building your Wasp app...')
+    cd(options.waspDir)
+    await $`${options.waspExe} build`
+  }
+
+  const tomlFiles = tomlHelpers.getTomlFileInfo(options)
+  let commonOps: CommonOps
+
+  if (options.context === SERVER_CONTEXT_OPTION) {
+    commonOps = new ServerCommonOps(options.waspDir, tomlFiles)
+  } else {
+    commonOps = new ClientCommonOps(options.waspDir, tomlFiles)
+  }
```

Comment:
Good call on ternary, and also using the enum and moving it into `CommonOps`, will do, thnx 👍🏻

Reply:
Nice but that still leave a bit of duplication, we can go the full way:

```ts
const commonOps =
  new (options.context === ServerContextOption.Server
       ? ServerCommonOps : ClientCommonOps) (options.waspDir, tomlFiles)
```

Hehe ok kidding a bit -> but I am interested now, would this work in JS?

Reply:
Haha, yes, it would work.

Funnily enough, it only works with constructors and when using the `new` keyword.

Assuming `doX` and `doY` are methods on `commonOps`, something like this:
```javascript
const commonOps =
  (options.context === ServerContextOption.Server
       ? commonOps.doX : commonOps.doY) (options.waspDir, tomlFiles)
```
Wouldn't work. More precisely, it would work, but not always how you'd expect (thanks to the binding of `this`).